src/stream/src/executor/backfill/arrangement_backfill.rs

```diff
@@ -0,0 +1,516 @@
+                    // but we still need to persist the finished state.
+                    // We currently persist it on the second barrier here rather than first.
+                    // This is because we can't update state table in first epoch,
+                    // since it expects to have been initialized in previous epoch
+                    // (there's no epoch before the first epoch).
+                    if is_snapshot_empty {
+                        current_pos =
+                            construct_initial_finished_state(pk_in_output_indices.len())
+                    }
+
+                    // We will update current_pos at least once,
+                    // since snapshot read has to be non-empty,
+                    // Or snapshot was empty and we construct a placeholder state.
+                    debug_assert_ne!(current_pos, None);
+
+                    persist_state(
+                        barrier.epoch,
+                        &mut self.state_table,
+                        true,
+                        &current_pos,
+                        &mut old_state,
+                        &mut current_state,
+                    )
+                    .await?;
+                    self.progress.finish(barrier.epoch.curr);
+                    yield msg;
+                    break;
+                }
+                yield msg;
+            }
+        }
+
+        // After progress finished + state persisted,
+        // we can forward messages directly to the downstream,
+        // as backfill is finished.
+        #[for_await]
+        for msg in upstream {
+            if let Some(msg) = mapping_message(msg?, &self.output_indices) {
+                if let Message::Barrier(barrier) = &msg {
+                    self.state_table.commit_no_data_expected(barrier.epoch);
+                }
+                yield msg;
+            }
+        }
+    }
+
+    /// Read snapshot per vnode.
+    /// These streams should be sorted in storage layer.
+    /// 1. Get row iterator / vnode.
+    /// 2. Merge it with futures unordered.
+    /// 3. Change it into a chunk iterator with `iter_chunks`.
+    /// This means it should fetch a row from each iterator to form a chunk.
+    /// Within each vnode, rows will be ordered by pk.
+    #[try_stream(ok = Option<StreamChunk>, error = StreamExecutorError)]
+    async fn snapshot_read_per_vnode<'a>(
+        schema: Arc<Schema>,
+        upstream_table: &'a ReplicatedStateTable<S>,
+        current_pos_map: &'a CurrentPosMap,
+    ) {
+        let mut streams = Vec::with_capacity(upstream_table.vnodes().len());
+        for vnode in upstream_table.vnodes().iter_vnodes() {
+            let current_pos = current_pos_map.get(&vnode);
+            let range_bounds = compute_bounds(upstream_table.pk_indices(), current_pos.cloned());
+            if range_bounds.is_none() {
+                continue;
+                // TODO: Highlight this change to reviewers.
+                // yield None;
```

Comment:
Previously we read from a single iterator, so we could just yield `None`.
Here we have multiple iterators. So we only yield None at the very end, and just ignore an iterator if it does not yield anything.